crates/rpc/src/lib.rs

```diff
@@ -83,66 +79,93 @@ impl RpcServer {
+            Hint: If this happens when upgrading, make sure to shut down the first one first.
+            Hint: If you are looking to run two instances of pathfinder, you must configure them with different http rpc addresses.", self.addr)),
         };
+        let addr = listener
+            .local_addr()
+            .context("Getting local address from listener")?;
+        let server = axum::Server::from_tcp(listener).context("Binding server to tcp listener")?;
+
+        async fn handle_middleware_errors(err: axum::BoxError) -> (http::StatusCode, String) {
+            use http::StatusCode;
+            if err.is::<tower::timeout::error::Elapsed>() {
+                (
+                    StatusCode::REQUEST_TIMEOUT,
+                    "Request took too long".to_string(),
+                )
+            } else {
+                // TODO: confirm this isn't too verbose.
+                tracing::warn!(error = err, "Unhandled middleware error");
+
+                (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    "Internal error".to_string(),
+                )
+            }
+        }
 
-        let methods = module.build();
+        let middleware = tower::ServiceBuilder::new()
+            // Convert errors created by middleware layers into responses.
+            // This is required by axum -- axum doesn't deal with Result, errors
+            // must be responses as well.
+            .layer(HandleErrorLayer::new(handle_middleware_errors))
+            .concurrency_limit(self.max_connections)
+            .layer(DefaultBodyLimit::max(REQUEST_MAX_SIZE))
+            .timeout(REQUEST_TIMEOUT)
+            .layer(tower_http::trace::TraceLayer::new_for_http())
+            .option_layer(self.cors);
+
+        /// Returns success for requests with an empty body without reading
+        /// the entire body.
+        async fn empty_body(request: Request<Body>) -> impl IntoResponse {
+            use hyper::body::HttpBody;
+            if request.body().is_end_stream() {
+                http::StatusCode::OK.into_response()
+            } else {
+                http::StatusCode::METHOD_NOT_ALLOWED.into_response()
+            }
+        }
+
+        let default_router = match self.default_version {
+            DefaultVersion::V03 => v03::rpc_router(),
+            DefaultVersion::V04 => v04::rpc_router(),
+        };
 
-        Ok(server.start(methods).map(|handle| (handle, local_addr))?)
+        let router = axum::Router::new()
+            // Also return success for get's with an empty body. These are often
+            // used by monitoring bots to check service health.
+            .route("/", get(empty_body).post(default_router))
+            .route("/rpc/v0.3", post(v03::rpc_router()))
+            .route("/rpc/v0.4", post(v04::rpc_router()))
+            .route("/rpc/pathfinder/v0.1", post(pathfinder::rpc_router()))
+            .layer(middleware)
+            // TODO: websockets
```

Comment:
It is definitely a breaking change.

We're okay with doing this since this is a pathfinder extension and not a starknet specification. Our existing implementation / specification for this is.. not great and should be re-done. 

Re-implementing the existing implementation will be a bit painful and probably useless -- we estimate very few / no-one is using this yet, as it only exposes headers and not even all the information required for them. I wanted to not delay this PR (its pretty large already); and rather delegate subscription API to a new PR with a consensus on the design and API for it.

I did already start with the bones of a subscription framework, but after thinking on it some, decided it was better to wait -- I did however leave the bones behind for reference. Which I should document as you point out elsewhere.